dali/test/python/test_pipeline.py

```diff
@@ -2333,3 +2333,32 @@ def def_ref():
         (ref,) = ref_pipe.run()
         check_batch(cpu, ref, bs, 0, 0, "HWC")
         check_batch(gpu, ref, bs, 0, 0, "HWC")
+
+
+def test_optional_build():
+    bs = 8
+
+    @pipeline_def(batch_size=bs, num_threads=4, device_id=0)
+    def pdef_regular():
+        enc, _ = fn.readers.file(file_root=jpeg_folder, name="only_reader")
+        img = fn.decoders.image(enc, device="mixed")
+        return img
+
+    @pipeline_def(batch_size=bs, num_threads=4, device_id=0)
+    def pdef_source():
+        source = fn.external_source(name="source")
+        return source
+
+    pipes = [pdef_regular() for _ in range(5)]
+
+    pipes[0].run()
+    pipes[1].schedule_run()
+    assert pipes[2].epoch_size("only_reader") != 0
+    assert pipes[3].executor_statistics() == {}
+    assert "shard_id" in pipes[4].reader_meta("only_reader")
+
+    pipes.append(pdef_source())
+    pipes[-1].feed_input("source", np.array([10, 10]))
+
+    for pipe in pipes:
+        assert pipe._built
```

Comment:
The name of the test suggests that it tests the (optional) build when in fact it tests _lack_ of build. This is clear now that we're used to having mandatory build, but will become a WTF for newcomers.
Moreover, if we believe that build should be optional, then adding a dedicated test for that seems superfluous - indeed, we should have a _reverse_ test (with manually called build) and just drop build from most of the tests. This would give us more coverage and possibly even discover some bugs by bringing build and run closer together.